application/src/main/java/org/thingsboard/server/service/computation/classloader/RuntimeJavaCompiler.java

```diff
@@ -0,0 +1,200 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.service.computation.classloader;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang.StringUtils;
+import org.thingsboard.server.service.computation.classloader.exception.CompilationException;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.jar.JarFile;
+import javax.tools.*;
+
+@Slf4j
+public class RuntimeJavaCompiler {
+    private static final String BANG = "!";
+    private static final String JAR = ".jar";
+    private StandardJavaFileManager fileManager;
+    private JavaCompiler compiler;
+    private File tempDir;
+    private final Object mutex = new Object();
+    private boolean classPathInitialized = false;
+
+    public RuntimeJavaCompiler(){
+        compiler = ToolProvider.getSystemJavaCompiler();
+        if (compiler == null) {
+            throw new IllegalStateException("Cannot find the system Java compiler. "
+                    + "Check that your class path includes tools.jar");
+        }
+        initTempDirectory();
+    }
+
+    private void initTempDirectory(){
+        try {
+            Path dir = Files.createTempDirectory("spark");
+            dir.toFile().deleteOnExit();
+            this.tempDir = dir.toFile();
+        } catch (IOException e) {
+            log.error("Error occurred while creating temp directory", e);
+        }
+    }
+
+    private void initFileManager(){
+        if(fileManager == null) {
+            StandardJavaFileManager fileManager = compiler.getStandardFileManager(null, null, null);
+            try {
+                fileManager.setLocation(StandardLocation.CLASS_OUTPUT, Arrays.asList(tempDir));
+                this.fileManager = fileManager;
+            } catch (Exception e) {
+                log.error("Error while setting class path", e);
+            }
+        }
+    }
+
+    private void initClassPath() throws IOException, URISyntaxException, CompilationException {
+        if(!classPathInitialized) {
+            Set<File> classPaths = new HashSet<>();
+            appendExistingClasspath(classPaths);
+            classPaths.add(tempDir);
+            ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+            if (contextClassLoader instanceof URLClassLoader) {
+                URLClassLoader load = (URLClassLoader) contextClassLoader;
+                addClassDirectoryToClassPath(load);
+                classPaths.addAll(parseUrlsAsClasspath(load.getURLs()));
+            }
+            fileManager.setLocation(StandardLocation.CLASS_PATH, classPaths);
+            classPathInitialized = true;
```

Comment:
As this is a System Java Compiler, it used System Classloader which doesn't have all the dependencies on Classpath. For that purpose i had to get Application Classloader with all the URLs on classpath and add them to FileManager used by JavaCompiler.
For local run with maven plugin it directly get absolute file paths as URL so it doesn't need extra processing. But in case of Production run using jar/debian/exe classpath URLs are nested like file:/user/share/thingsboard.jar!/BOOT-INF/org/apache/someJar.jar which can not be converted to Absolute file path unless we extract archive and then load it as file. For same purpose JarUnpacker is used to unpack jars and  to Parse URL from classloader to File parseUrlAsClasspath method is used.